### PR TITLE
add calc_growthrate function

### PR DESCRIPTION
**Pull request by orichters**
Calculates the growth rate of all variables in '%/yr' and returns a sorted data.frame where each variable name gets ` [Growth rate]` appended.

If, for example, your data contains the data in 2070 and 2060, the growth rate returned for 2070 is calculated as 100 * ((d2070/d2060)^(1/10) - 1). No growth rate can be calculated for the first year of the data. Infinite or undefined values (for example if d2060 = 0) are dropped.

Is the name ok or is there a risk of conflict with functions from other packages?

### CODE REVIEW

File: R/growthrate.R [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#' @return the sorted quitte object
+#' @export
+growthrate <- function(x) {
+  x %>%
+    as.quitte(na.rm = TRUE) %>%
+    quitteSort() %>%
+    group_by(.data$model, .data$scenario, .data$region, .data$variable) %>%
+    mutate(diffyear = .data$period - lag(.data$period)) %>%
+    mutate(growthrate = 100 * ((.data$value/lag(.data$value))^(1/.data$diffyear) - 1)) %>%
+    filter(! .data$growthrate %in% c(NA, Inf, NaN)) %>%
```
Comment (pweigmann) [inline R/growthrate.R:23]: just in case also include `-Inf` maybe?

**Comment by 0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q** [comment]
How about you figure out when you are done and request reviews then, so people do not have to play whack-a-mole …

File: R/growthrate.R [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#' @importFrom dplyr group_by mutate filter select rename
+#'
+#' @return the sorted quitte object
+#' @export
+growthrate <- function(x) {
+  x %>%
+    as.quitte(na.rm = TRUE) %>%
+    quitteSort() %>%
+    group_by(.data$model, .data$scenario, .data$region, .data$variable) %>%
+    mutate(diffyear = .data$period - lag(.data$period)) %>%
```
Comment (pweigmann) [inline R/growthrate.R:21]: Exists in multiple packages, better use via ``package::lag``?

**Comment by pweigmann** [comment]
the function is quite slow because of ``mutate(variable = factor(paste0(.data$variable, " [Growth Rate]")))``, maybe operate on levels later in the function

File: R/growthrate.R (hunk 2 above)
Comment (orichters) [reply R/growthrate.R]: I added it to the imports

File: R/growthrate.R (hunk 1 above)
Comment (orichters) [reply R/growthrate.R]: good idea

**Comment by orichters** [comment]
> How about you figure out when you are done and request reviews then, so people do not have to play whack-a-mole …

I agree, sorry, using the new IAMC style was an idea that came to my mind only later.

**Review by pweigmann — APPROVED** [review]

File: R/growthrate.R [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#' Calculates the growth rate of all variables in '%/yr' and returns a sorted data.frame
+#' where each variable name gets '|Growth rate' appended.
+#' If, for example, your data contains the data in 2070 and 2060, the growth rate returned
+#' for 2070 is calculated as 100 * ((d2070/d2060)^(1/10) - 1).
+#' No growth rate can be calculated for the first year of the data.
+#' Infinite or undefined values (for example if d2060 = 0) are dropped.
```
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [inline R/growthrate.R:6]: Bit of a long title.


image

File: R/growthrate.R [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+#'
+#' @param x A quitte object
```
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [inline R/growthrate.R:8]: Actually takes anything that has an `as.quitte()` method.  It is an undocumented feature people might come to rely on.

File: R/growthrate.R [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+#'
+#' @author Oliver Richters
+#'
+#' @importFrom dplyr filter group_by lag mutate rename select
+#'
+#' @return the sorted quitte object
+#' @export
+growthrate <- function(x) {
```
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [inline R/growthrate.R:16]: I would prefer are verb or verb & noun name for a function that transforms data, so e.g. `calc_growthrate()`.

File: R/growthrate.R [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  x <- as.quitte(x, na.rm = TRUE)
+  levels(x$variable) <- paste(levels(x$variable), "[Growth Rate]")
+  x %>%
+    quitteSort() %>%
```
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [inline R/growthrate.R:20]: Why is the data frame sorted?  To ensure that `lag()` uses the previous period?  That should be done explicitly using `lag(order_by = .data$period)`.

File: R/growthrate.R [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    group_by(.data$model, .data$scenario, .data$region, .data$variable) %>%
+    mutate(diffyear = .data$period - lag(.data$period),
+           growthrate = 100 * ((.data$value/lag(.data$value))^(1/.data$diffyear) - 1),
+           unit = factor("%/yr")) %>%
+    filter(! .data$growthrate %in% c(NA, -Inf, Inf, NaN)) %>%
```
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [inline R/growthrate.R:25]: `filter(is.finite(.data$growthrate))`

File: R/growthrate.R [hunk 8]
```diff
@@ -0,0 +1,29 @@
+#' Calculates the growth rate of all variables in '%/yr' and returns a sorted data.frame
+#' where each variable name gets '|Growth rate' appended.
```
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [inline R/growthrate.R:2]: > gets '|Growth rate' appended

That part is not true anymore.

**Review by 0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q — COMMENTED** [review]
While I understand this function is designed to fit the task at hand, I feel it would be easy to extend to make more useful for other applications.
Specifically, people might want to calculate growth rates only for a subset of variables and add them to their data frame for further processing.
They would have to do some 
```
bind_rows(
    x,

    x %>%
        filter(…) %>%
        calc_growthrate()
)
```
which could easily be wrapped in this function, with arguments for filtering variables and appending instead of replacing.

File: R/growthrate.R (hunk 3 above)
Comment (orichters) [reply R/growthrate.R]: Moved it into "details"

File: R/growthrate.R (hunk 4 above)
Comment (orichters) [reply R/growthrate.R]: adjusted it, also in quitteSort

File: R/growthrate.R (hunk 6 above)
Comment (orichters) [reply R/growthrate.R]: thanks, I adjusted it

File: R/calc_growthrate.R [hunk 9]
```diff
@@ -0,0 +1,30 @@
+#' Calculates the growth rate of all variables in '%/yr' and returns a quitte data.frame
```
[the remark this answers is not in the document]
Comment (orichters) [reply R/calc_growthrate.R]: I shortened it further

**Comment by orichters** [comment]
I implemented `only.new` as in `calc_addVariable` and `filter.function` as in `read.quitte`, with the special case that you can also simply pass a vector `c("GDP|PPP")` and it will be then use only those. Added some more tests and found some minor bugs (missing `ungroup`), but should be good to go now.

**Review by 0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q — APPROVED** [review]
Nice.

**Comment by 0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q** [comment]
This would actually be a good candidate to mention in the tutorial, @orichters.

**Comment by orichters** [comment]
quitte or remindmodel tutorial?

**Pull request merged**